src/USER-PINN/pair_gmu_pinn.cpp

```diff
@@ -0,0 +1,2251 @@
+}
+
+void PairGMUPINN::dcostheta_ijk_dri(double cstheta, double *drj,
+                                     double *drk, double *fj, double *fk)
+{
+  double rij, rik;
+
+  rij = sqrt(dot_product (drj, drj, 3));
+  rik = sqrt(dot_product (drk, drk, 3));
+
+  for (int i = 0; i < 3; i++) {
+    fj[i] = (-drk[i] / rik + cstheta * drj[i] / rij) / rij;
+    fk[i] = (-drj[i] / rij + cstheta * drk[i] / rik) / rik;
+  }
+}
+
+void PairGMUPINN::force_costheta_ijk_ri(double pref, double cstheta, double *drj, double *drk,
+                                        double *fi, double *fj, double *fk)
+{
+  double rij, rik;
+
+  rij = sqrt(dot_product (drj, drj, 3));
+  rik = sqrt(dot_product (drk, drk, 3));
+
+  for (int i = 0; i < 3; i++) {
+    fj[i] = pref * (-drk[i] / rik + cstheta * drj[i] / rij) / rij;
+    fk[i] = pref * (-drj[i] / rij + cstheta * drk[i] / rik) / rik;
+    fi[i] = fj[i] + fk[i];
+  }
+}
+
+double PairGMUPINN::dot_product(double *x, double *y, int n)
+{
+  double tmp = 0.0;
+
+  for (int i = 0; i < n; ++i) {
+    tmp += x[i] * y[i];
+  }
+
+  return tmp;
+}
+
+
+/* ---------------------------------------------------------------------- */
+
+double PairGMUPINN::bop_fa(int it, int jt, double r)
+{
+  // only -f_A(r)
+  return -exp(big_b[it][jt] - beta[it][jt] * r);
+}
+
+double PairGMUPINN::bop_fa_d(int it, int jt, Param *param, double r)
+{
+  // derivative of f_A(r)f_c(r) wrt 'r'
+
+  double tmp_fc, tmp_fc_d, tmp_exp;
+
+  tmp_fc = ann_fc (r, param);
+  tmp_fc_d = ann_fc_d (r, param);
+  tmp_exp = bop_fa (it, jt, r);
+
+  return tmp_exp * (tmp_fc_d - tmp_fc * beta[it][jt]) / r;
+}
+
+/* ---------------------------------------------------------------------- */
+
+double PairGMUPINN::bop_zeta(int it, int jt, int kt, Param *param,
+                             double s_ij, double cs_ijk, double r)
+{
+  double cut;
+
+  cut = ann_fc (r, param);
+  return s_ij * small_a[it][jt][kt] * cut * pow(cs_ijk - small_h[it][jt][kt], 2);
+}
+
+/* ---------------------------------------------------------------------- */
+
+double PairGMUPINN::screen_ijk(int it, int jt, int kt, Param *param, double r)
+{
+  return  1.0 - ann_fc (r, param) * exp(-lambda[it][jt][kt] * fabs(r));
+}
+
+/* ---------------------------------------------------------------------- */
+
+double PairGMUPINN::bop_fr(int itype, int jtype, double r)
+{
+  // only f_R(r) as in Tersoff
+
+  return exp(big_a[itype][jtype] - alpha[itype][jtype] * r);
+}
+
+/* ---------------------------------------------------------------------- */
+
+double PairGMUPINN::bop_fr_d(int itype, int jtype, Param *param, double r)
+{
+  // derivative of f_R(r)f_c(r) wrt 'r'
+
+  double tmp_exp, tmp_fc, tmp_fc_d;
+
+  tmp_fc = ann_fc (r, param);
+  tmp_fc_d = ann_fc_d (r, param);
+  tmp_exp = bop_fr (itype, jtype, r);
+
+  return tmp_exp * (tmp_fc_d - tmp_fc * alpha[itype][jtype]) / r;
+}
+
+
+/* ---------------------------------------------------------------------- */
+
+void PairGMUPINN::read_file(char *file)
```

Comment:
You may consider updating the file reading part using the `PotentialReader` class for more compact, thread safe, and buffer overflow safe reading/parsing for potential files, like it is done in several of the manybody potentials already.